src/main/scala/djinni/JNIGenerator.scala

```diff
@@ -35,6 +35,15 @@ class JNIGenerator(spec: Spec) extends Generator(spec) {
   def writeJniHppFile(name: String, origin: String, includes: Iterable[String], fwds: Iterable[String], f: IndentWriter => Unit, f2: IndentWriter => Unit = (w => {})) =
     writeHppFileGeneric(spec.jniHeaderOutFolder.get, spec.jniNamespace, spec.jniFileIdentStyle)(name, origin, includes, fwds, f, f2)
 
+  if(spec.jniGenerateMain) {
+    createFile(spec.jniOutFolder.get, "djinni_jni_main." + spec.cppExt, w => {
+      w.wl("// AUTOGENERATED FILE - DO NOT MODIFY!")
+      w.wl("// The file included here provides default JNI_OnLoad and JNI_OnUnload implementations.")
+      w.wl("// Disable the generation of this file with `--jni-generate-main=false`.")
+      w.wl(s"#include " + q(spec.jniBaseLibIncludePrefix + "djinni_jni_main.hpp"))
+    })
+  }
+
```

Comment:
This is where the magic happens. The file `djinni_jni_main.cpp` will be written to `--jni-out`. It contains nothing but an include of the header `#include djinni/jni/dijnni_jni_main.hpp` that implements `JNI_OnLoad` & `JNI_OnUnload`. This way, if the user builds a shared library from the generated sources, it will automatically have the correct ABI for use with JNI. 

This does not help if the user builds a static lib from the generated sources & links it from another shared library. In that case the user would have to include the header from the support-lib himself again. 

The good thing: It doesn't matter if this is part of a static libraries ABI. It will get stripped away when linked by another library anyways.

Reply:
I am not convinced that is needed, why not always let the user include `dijnni_jni_main.hpp` , if we choose to go that route .... in that way the usage is always the same, no matter it the support lib is used static or dynamic.
Or, just disable a dynamic support lib ... 

let us think more about the question, but I really would like to avoid additional command line arguments that we can/do not test.

Reply:
I am trying to solve the following: 

If the user would have to include the `djinni_jni_main.hpp` himself (which would be the easiest solution, I agree), the instructions would sound something like this:

"If building for JNI, include `djinni_jni_main.hpp` from the support lib somewhere in your target sources."

The question is: where should the header be included? In a normal djinni project setup all non-generated C++ code should be platform independent. So the user would have to

- add source file that includes `djinni_jni_main.hpp`, which only is added to the build sources if building for JNI or
- add a preprocessor condition that only includes the header if building for JNI.

This adds complexity to the users build configuration.

If we would decide to go a different direction and tell the user to include a sourcefile provided by the support lib that implements `JNI_OnLoad`, like you suggested in our slack discussion, the install instructions would be:

"If building for JNI, add `djinni_jni_main.cpp` to your target sources"

This comes with the same overhead. The user has to add logic to his build configuration that includes the file if he wants to build for JNI, otherwise not.

The new option `--jni-generate-main` moves that required complexity away from the user. Now he doesn't even have to know about the internals and the generated code will just work without him adding additional logic for it.

Reply:
we have 4 different scenarios
* static - our init
* static -custom init
* dynamic - our init
* dynamic - custom init
shall we put this into a own PR, where we might play all that 4 rules, and see how it looks like in real when doing so,
or shall we do it here?

Reply:
To reduce the complexity of this PR I plan to create separate smaller PRs for the stuff from this PR that we agree on and that doesn't need further discussion (documentation, improved tests). 
That way this PR will shrink in size once we merged the smaller changes. Does that make sense or am I missing sth?
In the end this PR should more or less only contain stuff regarding the newly introduced parameter. So let's discuss the new parameter here. 👍 

I'm not sure if I understand you 4 scenarios correctly:
1. **static support-lib** linked into shared library with **default init** *(-> when `--jni-generate-main=true`)*
2. **static support-lib** linked into shared library with **custom init** *(-> `--jni-generate-main=false` + custom implementation. Requires the user to write platform specific configuration logic.)*
3. **shared support-lib with default init** linked by shared library with **default init**
4. **shared support-lib with default init** linked by shared library with **custom init**
5. ~~dynamic support-lib with custom init linked by shared library~~ *(-> If building the support lib as dynamic library, it is self contained and only needs to initialize itself. Therefore no custom implementation is required.)*

Reply:
sounds good!
yes, 
I think 3 and 4 are highly unlikely use cases anyway, so we can/should focus on 1 and 2 
The default behaviour should be, I do nothing, have no custom init, and it just works. 
Is that possible and if not, what is the minimum action a user has to do?

Reply:
I tried to sketch the different variations.

1 & 3 require no knowledge of the internals if the proposed `--jni-generate-main` is introduced:

![djinni_jni (4)](https://user-images.githubusercontent.com/21294002/116795326-6d642380-aad4-11eb-9c71-0052735200a6.png)

Reply:
ok, now I think I understand the whole process, please let me know if I summaries that now correctly:

* djinni_jni_main.cpp get generated as all the other files from the IDL, 
* it will therefore be included in the list of generated files, 
* user will get it into users_shared.lib by just adding all generated files, as today
* the default setting is to generatedjinni_jni_main.cpp, so everything just works out of the box! 

If I got it now right, it makes a lot of sense.
Awesome as always, Johannes!
Thanks for your patience explaining me the new details!!

Reply:
Yes, that's correct!